### PR TITLE
chore: migrate to github-actions

### PR DESCRIPTION
**Pull request by hiroppy**
This PR contains a:

- [ ] **bugfix**
- [ ] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [x] **metadata update**

### Motivation / Use-Case

Migrate from azure-pipelines to github-actions

### Breaking Changes

no

### CODE REVIEW

**Comment by hiroppy** [comment]
I can't resolve this issue but a turning point is here. https://github.com/webpack/webpack-dev-middleware/commit/23a700ab01be3ff27c93419705ccc4b9c4f90565

**Comment by hiroppy** [comment]
Maybe this one. https://github.com/webpack/watchpack/issues/118

**Comment by evilebottnawi** [comment]
That error is happening when tests are failed and watch not closed

**Comment by hiroppy** [comment]
Yes, I know but when you skip https://github.com/webpack/webpack-dev-middleware/blob/master/test/middleware.test.js#L29, you will be able to see this error(https://github.com/webpack/webpack-dev-middleware/pull/611#issuecomment-608465254).

**Comment by hiroppy** [comment]
CI is green when I specify to comment out some places.

**Comment by hiroppy** [comment]
Actually I want to move to github CI:)

**Comment by hiroppy** [comment]
Anyway, I'll resolve this problem and then I'll refactor many things.

File: test/middleware.test.js [hunk 1]
```diff
@@ -370,7 +370,7 @@ describe('middleware', () => {
       });
     });
 
-    describe('should work with difference requests', () => {
+    describe.skip('should work with difference requests', () => {
```
Comment (hiroppy) [inline test/middleware.test.js:373]: This test causes.

**Comment by hiroppy** [comment]
When we use only MacOSnode-10-v4, CI will be succeed but when all CIs runs, MacOSnode-10-v4 will be failed 🤔

**Comment by hiroppy** [comment]
Added CODECOV_TOKEN to settings/secret.

**Comment by hiroppy** [comment]
windows CIs seem flaky, need to investigate.

**Comment by hiroppy** [comment]
I'll refactor tests after merging this pr.

**Comment by hiroppy** [comment]
CI is green.

**Comment by hiroppy** [comment]
I'll modify github's setting of branch settings.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        run: npm ci
+      - name: Install webpack@${{ matrix.webpack-version }}
+        run: npm i webpack@${{ matrix.webpack-version }}
+      - name: Run tests with coverage
+        run: npm run test:coverage -- --ci --reporters="default" --reporters="jest-junit"
+      - name: Submit coverage data to codecov
+        uses: codecov/codecov-action@v1.0.6
+        with:
+          token: ${{ secrets.CODECOV_TOKEN }}
+          file: '**/junit.xml'
```
Comment (evilebottnawi) [inline .github/workflows/tests.yml:41]: Let's union two files in one, they are small and can be combined

File: .github/workflows/tests.yml [hunk 3]
```diff
@@ -0,0 +1,41 @@
+name: test
+
+on: [push]
```
Comment (evilebottnawi) [inline .github/workflows/tests.yml:3]: on:
  push:
    branches:
      - master
      - next
  pull_request:
    branches:
      - master
      - next

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+
+jobs:
+  test:
+    runs-on: ${{ matrix.os }}
```
Comment (evilebottnawi) [inline .github/workflows/tests.yml:7]: Lint - ${{ matrix.os }} - Node v${{ matrix.node-version }}

File: .github/workflows/linter.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
+name: linter
+
+on: [push]
+
+env:
+  NODE_VERSION: 10.x
+
+jobs:
+  linter:
+    runs-on: ubuntu-latest
```
Comment (evilebottnawi) [inline .github/workflows/linter.yml:10]: name: Lint - ${{ matrix.os }} - Node v${{ matrix.node-version }}

Comment (evilebottnawi) [inline .github/workflows/linter.yml:10]: runs-on: ${{ matrix.os }}

    strategy:
      matrix:
        os: [ubuntu-latest]
        node-version: [12.x]

File: .github/workflows/linter.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+          node -v
+          npm -v
+      - name: Install dependencies
+        run: npm ci
+      - name: 'Run lint'
+        run: npm run lint
+      - name: 'Run NPM audit'
+        run: npm run security
+      - name: Check commit message
+        uses: wagoid/commitlint-github-action@v1
```
Comment (evilebottnawi) [inline .github/workflows/linter.yml:30]: :+1:

File: .github/workflows/tests.yml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    strategy:
+      matrix:
+        os: [macos-latest, windows-latest, ubuntu-latest]
+        node-version: [13.x, 12.x, 10.x]
+        webpack-version: [next, 4]
+        exclude:
+          - node-version: 13.x
+            webpack-version: 4
+          - node-version: 12.x
+            webpack-version: 4
```
Comment (evilebottnawi) [inline .github/workflows/tests.yml:17]: No need exclude

File: .github/workflows/tests.yml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+        run: npm i -g npm
+      - name: Print versions
+        run: |
+          node -v
+          npm -v
+      - name: Install dependencies
+        run: npm ci
+      - name: Install webpack@${{ matrix.webpack-version }}
+        run: npm i webpack@${{ matrix.webpack-version }}
+      - name: Run tests with coverage
```
Comment (evilebottnawi) [inline .github/workflows/tests.yml:35]: name: Run tests for webpack version ${{ matrix.webpack-version }}

File: .github/workflows/tests.yml [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+          node-version: ${{ matrix.node-version }}
+      - name: Use Npm@latest
+        run: npm i -g npm
+      - name: Print versions
+        run: |
+          node -v
+          npm -v
+      - name: Install dependencies
+        run: npm ci
+      - name: Install webpack@${{ matrix.webpack-version }}
```
Comment (evilebottnawi) [inline .github/workflows/tests.yml:33]: name: Install webpack ${{ matrix.webpack-version }}

File: .github/workflows/tests.yml (hunk 7 above)
Comment (hiroppy) [reply .github/workflows/tests.yml:17]: Do you know the `exclude` meaning? It is necessary.

File: .github/workflows/linter.yml (hunk 5 above)
Comment (hiroppy) [reply .github/workflows/linter.yml:10]: run-on is specified os version.

File: .github/workflows/tests.yml (hunk 3 above)
Comment (hiroppy) [reply .github/workflows/tests.yml:3]: why do you want to specify the two branches?

File: .github/workflows/linter.yml (hunk 5 above)
Comment (evilebottnawi) [reply .github/workflows/linter.yml:10]: Useful when you fast look on results, faced many times with this

File: .github/workflows/tests.yml (hunk 7 above)
Comment (evilebottnawi) [reply .github/workflows/tests.yml:17]: We need test what our middleware compatibility with webpack@4 + node@12 and webpack@4 + node@13

File: .github/workflows/linter.yml (hunk 5 above)
Comment (hiroppy) [reply .github/workflows/linter.yml:10]: sorry I can't understand. you can use the suggestion feature.

Comment (evilebottnawi) [reply .github/workflows/linter.yml:10]: Just add name for job :smile:

Comment (evilebottnawi) [reply .github/workflows/linter.yml:10]:
```
 name: Lint - ${{ matrix.os }} - Node v${{ matrix.node-version }}

runs-on: ${{ matrix.os }}
```

File: .github/workflows/tests.yml (hunk 7 above)
Comment (hiroppy) [reply .github/workflows/tests.yml:17]: Screen Shot 2020-04-06 at 21 50 17

So you want to change these tests cases?

File: .github/workflows/tests.yml (hunk 3 above)
Comment (evilebottnawi) [reply .github/workflows/tests.yml:3]: Because we should run on pull_request too, we always use next for next major release, so when we start to work the next time on a major release, we have to add this, we can avoid it and add it initially

File: .github/workflows/tests.yml (hunk 7 above)
Comment (evilebottnawi) [reply .github/workflows/tests.yml:17]: Yes, we should do it, some deps in theory can break webpack (for example inject globals in node runtime, it is very rare edge case, but better test it) , better to test it

**Comment by hiroppy** [comment]
Need to wait for resolving this issue. https://github.com/actions/setup-node/issues/132

**Comment by evilebottnawi** [comment]
github actions syntax is awful, no if/else, no ternary, a lot of examples broken from docs, I expected much better

**Comment by evilebottnawi** [comment]
Tests still broken on macos, investigate

**Comment by evilebottnawi** [comment]
Interesting: `line 1:  1597 Segmentation fault: 11  node_modules/.bin/jest test/middleware.test.js -u --verbose`

**Comment by hiroppy** [comment]
What's the Node version?

**Comment by evilebottnawi** [comment]
13

**Comment by evilebottnawi** [comment]
```
PID 3148 received SIGSEGV for address: 0x5c8
0   segfault-handler.node               0x00000001050a4066 _ZL16segfault_handleriP9__siginfoPv + 310
1   libsystem_platform.dylib            0x00007fff6ca445fd _sigtramp + 29
2   ???                                 0x0000000110a111ae 0x0 + 4573958574
3   node                                0x0000000100007fba _ZN4node16EmitAsyncDestroyEPN2v87IsolateENS_13async_contextE + 122
4   fse.node                            0x00000001050af59d _ZN3fse8FSEventsD2Ev + 93
5   fse.node                            0x00000001050af5fe _ZN3fse8FSEventsD0Ev + 14
6   node                                0x000000010032624f _ZN2v88internal13GlobalHandles28InvokeFirstPassWeakCallbacksINS1_4NodeEEEmPNSt3__16vectorINS4_4pairIPT_NS1_22PendingPhantomCallbackEEENS4_9allocatorISA_EEEE + 143
7   node                                0x00000001003261a3 _ZN2v88internal13GlobalHandles28InvokeFirstPassWeakCallbacksEv + 19
8   node                                0x000000010035d819 _ZN2v88internal4Heap24PerformGarbageCollectionENS0_16GarbageCollectorENS_15GCCallbackFlagsE + 2361
9   node                                0x000000010035b927 _ZN2v88internal4Heap14CollectGarbageENS0_15AllocationSpaceENS0_23GarbageCollectionReasonENS_15GCCallbackFlagsE + 1143
10  node                                0x0000000100367cda _ZN2v88internal4Heap33AllocateRawWithLightRetrySlowPathEiNS0_14AllocationTypeENS0_16AllocationOriginENS0_19AllocationAlignmentE + 58
11  node                                0x0000000100367d61 _ZN2v88internal4Heap34AllocateRawWithRetryOrFailSlowPathEiNS0_14AllocationTypeENS0_16AllocationOriginENS0_19AllocationAlignmentE + 33
12  node                                0x0000000100334c47 _ZN2v88internal7Factory15NewFillerObjectEibNS0_14AllocationTypeENS0_16AllocationOriginE + 87
13  node                                0x0000000100691928 _ZN2v88internal33Runtime_AllocateInYoungGenerationEiPmPNS0_7IsolateE + 168
14  node                                0x00000001009d4af9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit + 57
15  node                                0x00000001009d5935 Builtins_StringAdd_CheckNone + 661
16  node                                0x0000000100a38658 Builtins_AddHandler + 792
```

**Comment by evilebottnawi** [comment]
Problem in `FSEvents`

**Comment by evilebottnawi** [comment]
Very strange, because it sometimes doesn’t happen

**Comment by evilebottnawi** [comment]
I can not determine a problem(s) in node or jest :confused:

**Comment by evilebottnawi** [comment]
I even can't detect which tests lead to segment fault, look like `Math.random()`

**Comment by evilebottnawi** [comment]
2 tests passed, 3 - segment fault

**Comment by evilebottnawi** [comment]
Seems segment fault appears after `compiler = getCompiler(webpackMultiWarningConfig);` in 3rd tests

**Comment by evilebottnawi** [comment]
webpack@5 doesn't uses chokidar with fsevents and no problems, so I think the problem in `chokidar` or `fsevents`

File: src/index.js [hunk 10]
```diff
@@ -88,7 +88,8 @@ export default function wdm(compiler, options = {}) {
       // eslint-disable-next-line no-param-reassign
       callback = callback || noop;
 
-      context.watching.close(callback);
+      callback();
+      // context.watching.close(callback);
```
Comment (evilebottnawi) [inline src/index.js:92]: /cc @hiroppy The problem is here

Comment (hiroppy) [reply src/index.js:92]: Good, so can we resolve this problem?

Comment (evilebottnawi) [reply src/index.js:92]: I think bug in fsevents - https://github.com/fsevents/fsevents/issues/314, need to fix it, probably will have to figure it out yourself

Comment (evilebottnawi) [reply src/index.js:92]: Maybe related - https://github.com/webpack/webpack/issues/10037

File: .github/workflows/nodejs.yml [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    branches:
+      - master
+      - next
+  pull_request:
+    branches:
+      - master
+      - next
+
+jobs:
+  #  lint:
```
Comment (hiroppy) [inline .github/workflows/nodejs.yml:14]: Please uncomment lint job if possible

File: src/index.js (hunk 10 above)
Comment (hiroppy) [reply src/index.js:92]: Thank you for investigation!

File: .github/workflows/nodejs.yml (hunk 11 above)
Comment (evilebottnawi) [reply .github/workflows/nodejs.yml:14]: @hiroppy I will do it after we will fix the problem (before merge)

Comment (hiroppy) [reply .github/workflows/nodejs.yml:14]: okay, thanks

File: patch.js [hunk 12]
```diff
@@ -0,0 +1,755 @@
...
+// * paths     - string or array of strings, file/directory paths and/or globs
+// * options   - object, chokidar options
+
+// Returns an instance of FSWatcher for chaining.
+exports.watch = function(paths, options) {
+  return new FSWatcher(options).add(paths);
+};
+
+`
+);
```
Comment (evilebottnawi) [inline patch.js:755]: @hiroppy yes, problem(s) in `fsevents`, I patched `chokidar` does not use `fsevents` and tests passed.

Comment (paulmillr) [reply patch.js:755]: Now macos users would probably see webpack consuming 1GB of memory and tons of CPU.

Comment (evilebottnawi) [reply patch.js:755]: @paulmillr Firstly, this is not true, I can take measurements again and demonstrate them to you.  Secondly, I don’t see a desire to correct a bug or help us find it, so I don’t see any other choice how to turn off `fsevents`. I am always open and aimed at improvements and fix bugs, but for some reason you react negatively to my attempts to find a compromise.

Comment (paulmillr) [reply patch.js:755]: That's because we've already fixed the bug in chokidar 3 — and you should update. There is no "unwilingness", there is just desire to not do the same work twice.

Go ahead and remove fsevents. You'll notice some bug reports in a couple weeks.

Comment (evilebottnawi) [reply patch.js:755]: @paulmillr Why are you so negative to me, have I really done something bad to you? All I'm trying to do is find the error and fix it.

**Comment by hiroppy** [comment]
CI looks good

**Comment by evilebottnawi** [comment]
@hiroppy Let's merge :+1:

**Comment by hiroppy** [comment]
@evilebottnawi  Thank you for your investigation!

**Comment by hiroppy** [comment]
I'll copy to webapck-dev-server as well.
